nodl_python/nodl/types/_types.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2020 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it under the terms of the
+# GNU Limited General Public License version 3, as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranties of MERCHANTABILITY, SATISFACTORY QUALITY, or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Limited General Public License for more details.
+#
+# You should have received a copy of the GNU Limited General Public License along with
+# this program. If not, see <http://www.gnu.org/licenses/>.
+
+from abc import ABC
+from typing import Any, Dict, List
+
+from nodl._util import qos_to_dict
+from rclpy.qos import QoSPresetProfiles, QoSProfile
+
+
+class NoDLData(ABC):
+    """Data structure base class for NoDL objects."""
+
+    def __repr__(self) -> str:
+        return str(self._as_dict)
+
+    def __str__(self) -> str:
+        return str(self._as_dict)
+
+    @property
+    def _as_dict(self) -> Dict:
+        self_dict = self.__dict__.copy()
+        if 'qos' in self_dict:
+            self_dict['qos'] = qos_to_dict(self_dict['qos'])
+        return self_dict
+
+
+class NoDLInterface(NoDLData, ABC):
+    """Abstract base class for NoDL communication interfaces."""
```

Comment:
Ditto above: this isn't abstract, why not just use normal inheritance?

Reply:
97698c7